spark-operator/src/main/java/org/apache/spark/kubernetes/operator/utils/StatusRecorder.java

```diff
@@ -0,0 +1,201 @@
+ * Enables additional (extendable) observers for Spark App status.
+ * Cache & version locking might be removed in future version as batch app does not expect
+ * spec change after submitted.
+ */
+@Slf4j
+public class StatusRecorder {
+  protected final List<ApplicationStatusListener> appStatusListeners;
+  protected final ObjectMapper objectMapper = new ObjectMapper();
+  protected final ConcurrentHashMap<ResourceID, ObjectNode> statusCache;
+
+  public StatusRecorder(List<ApplicationStatusListener> appStatusListeners) {
+    this.appStatusListeners = appStatusListeners;
+    this.statusCache = new ConcurrentHashMap<>();
+  }
+
+  /**
+   * Update the status of the provided kubernetes resource on the k8s cluster. We use patch
+   * together with null resourceVersion to try to guarantee that the status update succeeds even
+   * if the underlying resource spec was update in the meantime. This is necessary for the correct
+   * operator behavior.
+   *
+   * @param resource Resource for which status update should be performed
+   */
+  @SneakyThrows
+  private void patchAndCacheStatus(SparkApplication resource, KubernetesClient client) {
+    ObjectNode newStatusNode =
+        objectMapper.convertValue(resource.getStatus(), ObjectNode.class);
+    ResourceID resourceId = ResourceID.fromResource(resource);
+    ObjectNode previousStatusNode = statusCache.get(resourceId);
+
+    if (newStatusNode.equals(previousStatusNode)) {
+      log.debug("No status change.");
+      return;
+    }
+
+    ApplicationStatus prevStatus =
+        objectMapper.convertValue(previousStatusNode, ApplicationStatus.class);
+
+    Exception err = null;
+    for (long i = 0; i < StatusPatchMaxRetry.getValue(); i++) {
+      // We retry the status update 3 times to avoid some intermittent connectivity errors
+      try {
+        replaceStatus(resource, prevStatus, client);
+        err = null;
+      } catch (KubernetesClientException e) {
+        log.error("Error while patching status, retrying {}/3...", (i + 1), e);
+        Thread.sleep(
+            TimeUnit.SECONDS.toMillis(StatusPatchFailureBackoffSeconds.getValue()));
+        err = e;
+      }
+    }
+
+    if (err != null) {
+      throw err;
+    }
+
+    statusCache.put(resourceId, newStatusNode);
+    appStatusListeners.forEach(listener -> {
+      listener.listenStatus(resource, prevStatus, resource.getStatus());
+    });
+  }
+
+  public void persistStatus(SparkApplicationContext context,
+                            ApplicationStatus newStatus) {
+    context.getSparkApplication().setStatus(newStatus);
+    patchAndCacheStatus(context.getSparkApplication(), context.getClient());
+  }
+
+  private void replaceStatus(SparkApplication resource, ApplicationStatus prevStatus,
+                             KubernetesClient client)
+      throws JsonProcessingException {
+    int retries = 0;
+    while (true) {
+      try {
+        var updated = client.resource(resource).lockResourceVersion().updateStatus();
```

Comment:
Just a very generic note about this, you might be already aware. There is no consensus operators where to store state. An other approach would be to add it to a `config map`/`secreate`/`custom resource`. In dependent resources we took rather this approach since that we were able to do in a loosely coupled way:
https://javaoperatorsdk.io/docs/dependent-resources#external-state-tracking-dependent-resources